architecture/2024-08-controller-component-threat-model.md

```diff
@@ -0,0 +1,241 @@
+- **Namespaces within the Cluster**: Logical partitions within the cluster to separate and isolate resources and workloads.
+
+The Controller component lives inside the `radius-system` namespace in the Kubernetes cluster where it is installed. UCPD also resides within the same namespace.
+
+The Kubernetes API Server, which is the main interactor of the Controller component, runs in the `kube-system` namespace within the cluster.
+
+### Key Points
+
+1. **Namespace Isolation**: Different namespaces separate and isolate resources and workloads within the Kubernetes cluster.
+2. **Access Controls**: Access controls and other permissions are implemented to manage interactions between namespaces.
+
+## Data Flow
+
+### Diagram
+
+![Controller Component via Microsoft Threat Modeling Tool](./2024-08-controller-component-threat-model/controller-component.png)
+
+1. **Resource Creation/Update/Delete**: When a user requests to create, update, or delete a Recipe or Deployment object, the request is sent to the Kubernetes API Server. For example, this request can be made by a human interactor using the `kubectl` command. The Kubernetes API Server authenticates and authorizes the request before the admission controllers intercept it.
+
+2. **Validating Webhook**: The only type of admission controller we have in Radius is the validating webhook for the Recipe resource. The validating webhook ensures that the Recipe object is one of the Radius portable resources. If the operation and the object is validated, then it is persisted to the **etcd**. You can see the details of the Recipe Webhook implementation via [this link](https://github.com/radius-project/radius/blob/main/pkg/controller/reconciler/recipe_webhook.go#L46).
+
+3. **Controller Reconciliation**: When the action (create/update/delete) is validated by the Validating Webhook (applicable only to the Recipe resource and not to the Deployment resource) as discussed in the previous step, the Kubernetes controllers we have (Recipe and Deployment Reconcilers) watch for changes via the Kubernetes API Server. The controllers then reconcile the resource's state to match the desired configuration specified in the resource definition. This involves ensuring that the actual state of the cluster matches the desired state, making any necessary adjustments to achieve this.
+
+4. **Communication with UCPD**: Controller also calls UCPD when doing reconcilations. Here is the list of instances where Controller calls UCPD:
+
+   1. Recipe Reconciler:
+
+      - Poll operations like create, update, or delete a Recipe resource.
+      - Resolve dependencies for a Recipe resource:
+        - Get the Radius environment the Recipe is associated with.
+        - Create the Radius Resource Group if needed.
+        - Create the Radius Application if needed.
+      - Fetch a Radius resource.
+      - List Secrets of a Recipe resource.
+      - Delete a Recipe resource.
```

Comment:
Should we give an example? I think this is technically accurate but still hard to grasp.